crates/network/README.md

```diff
@@ -0,0 +1,302 @@
+### Meta Protocols
+
+#### Identify
+
+- Purpose: Exchange of node metadata and supported protocols
+- Key Events:
+  - `Received`: Received identify information from a peer
+
+#### Ping
+
+- Purpose: Keep-alive and latency measurements
+
+### Custom Protocol
+
+#### Stream
+
+- Purpose: Application-specific data exchange
+- Implementation: Uses a custom `MessageJsonCodec` for serialization/deserialization
+- Features:
+  - Bidirectional communication between peers
+  - Message Structure: The protocol uses a custom `Message` struct defined as
+  
+ ```rust
+  pub  struct  Message  { pub data:  Vec<u8> }
+```
+
+This flexible structure allows for sending arbitrary data between peers.
+
+- Key Events:
+  - Handled directly in the `EventLoop` as `StreamOpened` events
+
+The custom stream protocol allows for tailored communication patterns specific to our application needs, complementing the standardized libp2p protocols.
+
+## Discovery
+
+We combine multiple protocols in order to improve our connectivity.
+
+- At the begging we dial Calimero boot nodes (which speak Calimero KAD protocol) and we start discovering peers on local network with mDNS.
+
+- During identify exchange (which occurs for every established connection) we record protocols which other peer supports.
+
+- For discovered peers with mDNS we perform direct dial (because we want to be connected to all local peers).
+
+- For discovered peers with rendezvous we perform dial only if peer is not already connected (this means that it most likely already discovered and dialed due to mDNS event).
+
+Beside network state by lip2p (more in `self.swarm.network_info()`) we also have our discovery state.
+
+We keep track of multiaddrs for all connected peers and peers of interest (which are never removed from state so we can reconnect to them when needed).
+
+To improve connectivity, as we discover relay nodes we attempt to make a relay reservation (so other peers can hole punch us).
+
+To improve connectivity, as we discover rendezvous nodes we attempt to register our external addrs (relayed addrs) and we attempt to discover other peers. Additionally, we periodically perform discovery against all discovered rendezvous nodes. At the moment we use single namespace for the rendezvous (check config), but we could use contexId for namespace.
+
+### Client (`NetworkClient` struct)
+
+The NetworkClient serves as the primary interface for other parts of the application to interact with the networking functionality. It provides a clean, easy-to-use API for performing various network operations without directly interacting with the internals of the EventLoop or libp2p.
+
+Key features of the NetworkClient include:
+
+- Simplified methods for common network operations (e.g., listening, dialing, subscribing to topics)
+- Abstraction of the underlying network complexity
+- Asynchronous operation support
+
+The NetworkClient extensively uses the oneshot pattern for most of its methods. This pattern involves:
```

Comment:
> oneshot pattern

is there such a thing? 😅 

this model is our close attempt at the actor pattern for lock-free state mutation without mandating large-scale refactors